### PR TITLE
an example of a word-level language model using FluxML

### PR DESCRIPTION
**Pull request by jaffourt**

### CODE REVIEW

File: text/word-rnn/word-rnn.jl [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                # compute loss for this batch
+                training_loss = loss(batch...)
+                return training_loss
+            end
+
+            for x in ps
+                # apply clip to handle exploding gradients
+                grad_x = clamp!(gradient[x], -args.clip, args.clip)
+                # backprop
+                x .-= lr .* grad_x
```
Comment (ToucheSir) [inline text/word-rnn/word-rnn.jl:82]: In the spirit of promoting best practices, we have https://fluxml.ai/Flux.jl/stable/training/optimisers/#Flux.Optimise.ClipValue and the rest of https://fluxml.ai/Flux.jl/stable/training/optimisers/ for this. I imagine you'd want to use something more sophisticated than plain SGD anyhow.

File: text/word-rnn/word-rnn.jl [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        end
+    end
+
+    #= For a more simple training loop:
+    # training callback for loss on hold-out set
+    evalcb() = @show(mean([loss(x_v, y_v) for (x_v, y_v) in hold_out]))
+    throttled_cb = Flux.throttle(evalcb, 5)
+    opt = ADAM(args.γ)
+    @Flux.epochs args.epochs Flux.train!(loss, ps, data_loader, opt, cb=throttled_cb)
+    =#
```
Comment (ToucheSir) [inline text/word-rnn/word-rnn.jl:104]:
```suggestion
```

File: text/word-rnn/word-rnn.jl [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+    # reference to model params, and optimizer
+    ps = Flux.params(model)
+
+    # create batch iterators for data and validation
+    data_loader = zip(x_train[1:end-5], y_train[1:end-5])
+    hold_out = zip(x_train[end-5:end], y_train[end-5:end])
+
+    # used for updating hyperparameters
+    best_val_loss = nothing
```
Comment (ToucheSir) [inline text/word-rnn/word-rnn.jl:61]:
```suggestion
    local best_val_loss
```

File: text/word-rnn/word-rnn.jl [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    # load train data and vocabulary
+    x_train, y_train, word2ind, vocab = get_data(args)
+    vocab_size = length(vocab)
+
+    # create model
+    model = create_model(vocab_size, args; kws...)
+
+    # logit cross entropy loss function
+    function loss(x, y)
+        Flux.reset!(model)
```
[the remark this answers is not in the document]
Comment (jaffourt) [reply text/word-rnn/word-rnn.jl:49]: @ToucheSir Can you elaborate on this? My understanding is that the loss function is called for each batch in a single training loop, thus for a sequential language model such as this one we want to reset the hidden state after each batch.

File: text/word-rnn/word-rnn.jl (hunk 1 above)
Comment (jaffourt) [reply text/word-rnn/word-rnn.jl:82]: @ToucheSir I opted out of the library methods as this explicit calculation provided better performance. However, for the promotion of best practices I agree and I can switch to using the library methods.
